### DISCUSSION TITLE
Deploying Nextjs app + storybook

### DISCUSSION THREAD

**Discussion by ribeiroguilherme** [vercel/next.js#14031] [Help]
Hey guys, I am trying to deploy my Nextjs app + its storybook together. I am having problems in serving the storybook static app inside my app.

What I have tried so far in my package json:
```json
{
  "scripts": {
    "build": "next build && next export && npm run build-storybook",
    "build-storybook": "build-storybook -c .storybook -o out/shared-ui",
    "deploy:staging": "now",
  }
}
```

Running the `build` generates the static version of my next app + storybook inside the `out` folder, creating a `shared-ui` that has all the storybook static files. If I run `npx http-server out` , everything works fine and I can access my storybook on `http://localhost:8080/shared-ui/`. 

When I deploy it and I try to access it through to the server, I am getting 404 on the Storybook inner iframe. I configured the `<base id="storybook-base-url" href="/shared-ui/"/>` inside the Storybook page and Iframe page, but nothing changed. Here is the URL:
https://chillout-fj99e6wrb.now.sh/shared-ui?path=/story/*

Any thoughts?

**Comment by kylemh** [comment]
I'm having the same problem. https://storybook-one-inky.vercel.app/?path=/story/*

Just gave the monorepo support a try and the inner storybook iframe 404s

**Reply by kylemh** [reply]
LOL I previously had added artifacts to `.vercelignore` which Storybook requires. Once I removed those items from `.vercelignore`, everything began to work again!

**Reply by leerob** [reply]
I'd love to see an example of this now that it's working 😄  I feel it's probably somewhat common.

**Reply by kylemh** [reply]
An example of Storybook with Next.js or the monorepo?

My example is open source: https://github.com/operationcode/front-end

**Reply by leerob** [reply]
Monorepo! I just realized the `with-storybook` example should suffice, just needs better docs 👍

**Comment by wagerfield** [comment]
What is the recommended approach for rendering components that use `next/link` in **Storybook** with `next@9.5.3`?

If you update the `HelloWorld` component in the `with-storybook` example to import `next/link` and render a `Link` you get the following error in the console:

```
link.js:7 Uncaught TypeError: Cannot read property 'prefetch' of null
    at prefetch (link.js:7)
    at link.js:22
    at link.js:3
    at Array.forEach (<anonymous>)
    at IntersectionObserver.rootMargin (link.js:3)
```

Here's the updated component in `with-storybook/components/index.js`

```js
import Link from "next/link"

export default function Home() {
  return (
    <Link href="/about">
      <a>About</a>
    </Link>
  )
}
```

I've Googled the error in question and found some [_stale solutions_](https://gist.github.com/novascreen/f1c44ead31e5a494556793be2c408840) for mocking `Router.router` in the `.storybook/preview.js` ...but that no longer works:

```js
// .storybook/preview.js
import Router from "next/router"

Router.router = {
  push: () => {},
  prefetch: () => Promise.resolve()
}
```

I've also found this [open issue](https://github.com/vercel/next.js/issues/16864) which is reporting the same error, but for a different reason.

The issue seems to be related to the fact that the `router` is not available in the context, so I would like to know what the _new solution_ is 😄 

I'm guessing the two obvious solutions are either:

1. Mock `next/router` using a new technique since the one above not longer seems to work
2. Create the Router context within Storybook using a global decorator in `.storybook/preview.js`

@Timer since you put together the `with-storybook` example, I'm tagging you here...and @arunoda since you provided the mock Router example, I'm tagging you too 👋 

Appreciate any help on this!

**Reply by kylemh** [reply]
You get the error in the console, but it still renders. I'm not sure if there's a huge issue with that for Storybook.

**Reply by wagerfield** [reply]
@kylemh thank you for the swift response!

I ended up replacing the implementation of `next/link` via [webpack's `NormalModuleReplacementPlugin`](https://webpack.js.org/plugins/normal-module-replacement-plugin/):

```js
// .storybook/main.js
const { resolve } = require("path")
const { NormalModuleReplacementPlugin } = require("webpack")

module.exports = {
  stories: ["../src/**/*.stories.@(mdx|tsx)"],
  addons: ["@storybook/addon-essentials"],
  webpackFinal(config) {
    config.plugins.push(
      new NormalModuleReplacementPlugin(
        /next\/link/,
        resolve(__dirname, "next-link.js")
      )
    )

    return config
  },
}
```

...and then mocking/replacing the implementation like so:

```js
// .storybook/next-link.js
const { createElement } = require("react")
const { action } = require("@storybook/addon-actions")

function onClick(event) {
  event.preventDefault()
  action("onClickLink")(event.target.href)
}

module.exports = ({ children, href }) =>
  createElement("a", { ...children.props, href, onClick })
```

This has a few benefits:

1. It reduces the size of Storybook's JS marginally since it's no longer bundling `next/link`
2. It removes the error from the console (as harmless as it may be)
3. It logs the `href` of the `Link` to Storybook's Actions panel (screenshot below)

<img width="461" alt="Screenshot 2020-10-08 at 09 03 26" src="https://user-images.githubusercontent.com/412747/95431519-2ef2f580-0945-11eb-9902-a36b7b540e16.png">

Hope that's helpful for anyone else using Storybook with Next.js 🎉

**Reply by DeadDuck83** [reply]
@wagerfield Thank you for the very clear and organized solution for the links. 

Would you be able to provide a suggestion for a deployed build to GitLab Pages? 
(also please forgive me, I am new to SSR and Storybook 😬)

Currently I have been Googling around and created my Pages yml file:
```yaml
pages:
  stage: deploy
  script:
  - yarn
  - yarn run build-storybook
  - mv storybook-static/* .public
  artifacts:
    paths:
    - public
  only:
  - master
```
That provides most of the functionality I am looking for, however, I use `Rosetta` as a translator and all of the app-level translated content does not get translated in the build:

![example](https://user-images.githubusercontent.com/44219378/96619474-1676c980-12bb-11eb-849b-10feab19e83b.jpg)

Could you possibly think of a solution for this?

This is what I am currently doing with the deploy files:
<img width="544" alt="Screen Shot 2020-10-20 at 10 02 32 AM" src="https://user-images.githubusercontent.com/44219378/96619795-71102580-12bb-11eb-8f0d-0c5bcce69960.png">

I am going to try to set up the Build on a branch directly on the Dev Repo to see if that will resolve the conflict, however, I would like to have the repo separate if possible.

**Comment by shpsyte** [comment]
> Hey guys, I am trying to deploy my Nextjs app + its storybook together. I am having problems in serving the storybook static app inside my app.
> 
> What I have tried so far in my package json:
> 
> ```json
> {
>   "scripts": {
>     "build": "next build && next export && npm run build-storybook",
>     "build-storybook": "build-storybook -c .storybook -o out/shared-ui",
>     "deploy:staging": "now",
>   }
> }
> ```
> 
> Running the `build` generates the static version of my next app + storybook inside the `out` folder, creating a `shared-ui` that has all the storybook static files. If I run `npx http-server out` , everything works fine and I can access my storybook on `http://localhost:8080/shared-ui/`.
> 
> When I deploy it and I try to access it through to the server, I am getting 404 on the Storybook inner iframe. I configured the `<base id="storybook-base-url" href="/shared-ui/"/>` inside the Storybook page and Iframe page, but nothing changed. Here is the URL: https://chillout-fj99e6wrb.now.sh/shared-ui?path=/story/*
> 
> Any thoughts?

Did you find a solution?

**Comment by leerob** [comment]
For folks looking to deploy both a Next.js site and a Storybook from the same folder to Vercel, we now have a guide: https://vercel.com/guides/storybook-with-vercel

**Reply by shpsyte** [reply]
> For folks looking to deploy both a Next.js site and a Storybook from the same folder to Vercel, we now have a guide: https://vercel.com/guides/storybook-with-vercel

But it will deploy only the storybook right? I'm trying to deploy both a Next.js site and a Storybook

Following the guide, I'll have only the SB published

**Reply by shpsyte** [reply]
Not sure if it is possible actually, better to have deploys separately using GitHub action  ...

**Reply by leerob** [reply]
No, you should be able to have multiple projects deployed from the same repository.

**Reply by Alarid** [reply]
> No, you should be able to have multiple projects deployed from the same repository.

Could you please elaborate ? I'm in the same situation as @shpsyte, and after following the guide I only have the storybook deployed. Not sure how to have both deployed

**Reply by davidtranjs** [reply]
Me too, same need with @Alarid and @shpsyte

**Reply by leerob** [reply]
You can deploy multiple projects from the same git repo. So there's two separate Vercel projects, one for you Next.js site and one for Storybook. Currently, you can't deploy _both_ at the same time (we're working on this). You have to deploy them one after another.

**Reply by OgDev-01** [reply]
How can we deploy them one after the the other @leerob ?

**Reply by philwolstenholme** [reply]
@leerob I think what I'm missing from that page (and maybe the others above) is how the `vercel.json` config doesn't mess up my deployment of the Next app? It looks like the presence of that file as described on https://vercel.com/guides/storybook-with-vercel would mean only the Storybook would be deployed as each time `vercel.json` was read it would use the Storybook output directory:

```
 "outputDirectory": "./storybook-static"
 ```
 
 Or is there something that I'm missing?
 
 I have one repo and am happy to do one deploy then another in my CI, but not sure how to make one deploy use the project default and the subsequent deploy use the overriden settings from that `vercel.json` example. Other than write to that file between the two deploys?
 
Edit: I'm using https://github.com/vercel/examples/tree/main/ci-cd/gitlab-cicd which probably doesn't help, as that has the project ID as an environment variable so I'm not sure how I'd deploy a separate Storybook project via Gitlab CI 🤔

**Reply by philwolstenholme** [reply]
Maybe I could use a combination of setting the `VERCEL_PROJECT_ID` env var to be my separate Storybook project as per https://vercel.com/guides/using-vercel-cli-for-custom-workflows and then using the `--local-config` argument in the Vercel CLI to point at a Storybook-specific version of a `vercel.json` file called something like `vercel.storybook.json` with the right build command, output directory etc. Then run the build and deploy commands?

**Reply by qq99** [reply]
@philwolstenholme That's my take from reading that guide as well, it seems like it'd be dangerous to my main project & take it offline, so I didn't attempt to run through that guide.

**Reply by qq99** [reply]
I achieved the deployment by making a new project using vercel dashboard UI, then using the UI to override the variables in the manner described at https://vercel.com/guides/storybook-with-vercel#deploying-storybook-to-vercel. So, it's doable without the vercel.json in your git repo

**Reply by y-a-v-a** [reply]
The issue with the example repo is that the stories are in a separate directory. This goes against Storybook's best practice of keeping stories next to components (next to `*.test.tsx`, next to `*.hooks.tsx`, etc.). But it seems the only solution.

When placing `*.stores.tsx` next to my components, `@storybook/react` is needed as a `dependency` (contrary to what Storybook is: a `devDependencies` entry, imho) and `next build` does type checks so it needs `@storybook/react`. In my Docker image, I don't want `devDependencies` as the image will become a production container.

The solution I will opt for is to disable (though discouraged) type checking during `next build` via `ignoreBuildErrors: true`.

**Reply by bscaspar** [reply]
According to #1321, Storybook is now a framework setting with Vercel CLI 29.1.1. 

Does this simplify this process at all? I am also confused about deploying both a next and Storybook project from the same repo. Would this require turbopack or something similar that supports a monorepo?

**Comment by gutisalex** [comment]
vercel cli is pulling the project id from the environment variables in gitlab / github... 

So you basically have two projects in vercel, one for the app and one for storybook. Then you create two vercel config files in your root folder name them whatever you like (ie. vercel-storybook.json and vercel-app.json)

.vercel-app.json
```json
{
  "$schema": "https://openapi.vercel.sh/vercel.json",
  "buildCommand": null,
  "devCommand": null,
  "installCommand": null,
  "framework": "nextjs",
  "outputDirectory": null
}
```

vercel-storybook.json
```json
{
  "$schema": "https://openapi.vercel.sh/vercel.json",
  "buildCommand": "npm run build-storybook",
  "devCommand": "npm run storybook",
  "installCommand": "npm install",
  "framework": null,
  "outputDirectory": "./storybook-static"
}
```

Now login to your account check the settings tab in both projects and copy the project ids. Create two env variables in your git lab ci/cd settings and paste the ids in there. **Don't** use the key VERCEL_PROJECT_ID. Instead use VERCEL_APP_PROJECT_ID and VERCEL_STORYBOOK_PROJECT_ID.

Now in your .gitlab-ci.yaml you have something like this:

```yaml
default:
  image: node:16.16.0

deploy_preview:
  stage: deploy
  except:
    - main
  script:
    - npm install --global vercel
    - vercel pull --yes --environment=preview --token=$VERCEL_TOKEN
    - vercel build --token=$VERCEL_TOKEN
    - vercel deploy --prebuilt  --token=$VERCEL_TOKEN
  variables:
    VERCEL_PROJECT_ID: $VERCEL_APP_PROJECT_ID

deploy_vercel_storybook:
  stage: deploy_storybook
  except:
    - main
  script:
    - npm install --global vercel
    - vercel pull --yes --environment=preview --token=$VERCEL_TOKEN
    - vercel build --token=$VERCEL_TOKEN --local-config vercel-storybook.json
    - vercel deploy --prebuilt --token=$VERCEL_TOKEN --local-config vercel-storybook.json
  variables:
    VERCEL_PROJECT_ID: $VERCEL_STORYBOOK_PROJECT_ID

deploy_production:
  stage: deploy
  only:
    - main
  script:
    - npm install --global vercel
    - vercel pull --yes --environment=production --token=$VERCEL_TOKEN
    - vercel build --prod --token=$VERCEL_TOKEN
    - vercel deploy --prebuilt --prod --token=$VERCEL_TOKEN
  variables:
    VERCEL_PROJECT_ID: $VERCEL_APP_PROJECT_ID
````

**Comment by asnaseer** [comment]
We have a nextjs app named `web` which is built in vercel. Everything builds and deploys fine.

We recently added storybook to this which added the following scripts to our `package.json` file:
```
    "storybook": "storybook dev -p 6006",
    "build-storybook": "storybook build"
```
When running storybook locally via `yarn storybook` everything looks fine. If we build the storybook locally via `yarn build-storybook` then a static site is created under the folder `storybook-static`. If we look at the JS files within this folder we can see that the storybook build correctly replaced all `process.env.XYZ` variables with their actual values which are defined within `.env.local` file.

We then created a new vercel project naming it `web-storybook` and setting the `Framework Preset` to `Storybook`. We left the `Build and Output Settings` at their default values and added the environment variables used by our stories.

Vercel built and deployed this however, it looks like it did not inject the environment variables into the build files under `storybook-static`. As an example, one of these environment variables is named `NEXT_PUBLIC_TWICPICS_DOMAIN`.

Both the `web` and `web-storybook` vercel projects point to the same GitHub source.

Would anyone here know why storybook is failing to inject the environment variables when built under vercel?
